python/ray/monitor.py

```diff
@@ -488,6 +492,106 @@ def driver_removed_handler(self, unused_channel, data):
 
         self._clean_up_entries_for_driver(driver_id)
 
+    def _xray_clean_up_entries_for_driver(self, driver_id):
+        """Remove this driver's object/task entries from redis.
+
+        Removes control-state entries of all tasks and task return
+        objects belonging to the driver.
+
+        Args:
+            driver_id: The driver id.
+        """
+
+        xray_task_table_prefix = (
+            ray.gcs_utils.TablePrefix_RAYLET_TASK_string.encode("ascii"))
+        xray_object_table_prefix = (
+            ray.gcs_utils.TablePrefix_OBJECT_string.encode("ascii"))
+
+        task_table_infos = {}  # task id -> TaskInfo
+        task_table_objects = self.state.task_table()
+        driver_id_hex = binary_to_hex(driver_id)
+        for task_id_hex in task_table_objects:
+            if len(task_table_objects[task_id_hex]) == 0:
+                continue
+            task_table_object = task_table_objects[task_id_hex][0]['TaskSpec']
+            task_driver_id_hex = task_table_object['DriverID']
+            if driver_id_hex != task_driver_id_hex:
+                # Ignore tasks that aren't from this driver.
+                continue
+            task_table_infos[task_id_hex] = task_table_object
+        driver_task_id_bins = [
+            hex_to_binary(task_id_hex)
+            for task_id_hex in task_table_infos.keys()
+        ]
+
+        # Get the list of objects returned by driver tasks.
+        driver_object_ids = set()
+        for task_info in task_table_infos.values():
+            driver_object_ids |= {
+                object_id.id()
+                for object_id in task_info['ReturnObjectIDs']
+            }
+
+        # Also record all the ray.put()'d objects.
+        all_put_objects = []
+        object_table_objects = self.state.object_table()
+        for object_id in object_table_objects:
+            if len(object_table_objects[object_id]) == 0:
+                continue
+            object_id_bin = object_id.id()
+            task_id_bin = ray.local_scheduler.compute_task_id(object_id).id()
+            all_put_objects.append((object_id_bin, task_id_bin))
+
+        # Keep objects from relevant tasks.
+        driver_task_ids_set = set(driver_task_id_bins)
+        for object_id, task_id in all_put_objects:
```

Comment:
If this actually covers all objects and not just put objects, then maybe we don't need to loop through the task return values above, is that right?